README.md
# Base2

Base2 is an Elixir library for encoding and decoding Base2 binaries. 

Is this a problem worth solving? Yes, but only if you need to work with Base2 as actual binaries, not just bits.

Working with Base2 is a relatively simple task chaining a few Elixir built-in functions or using a third-party generic "BaseX" type library, but most of these implementations leave a lot to be desired. Generally, most built-in methods and third-party libraries are often not very optimized. Using built-in functions also is not uniform with other ways of handling Bases such as via the Elixir `Base` module. Most of these methods are great for scratch work, but less suitable for bulk encoding and decoding and working at scale. Further, the multiple ways of approaching different bases lead to very inconsistent interfaces, for instance `Integer.to_string()` vs. `Base` vs a third-party module with its own conventions.

`Base2` includes the following functionality:

* Encodes and Decodes binaries as Base2.
* Consistent with the `Base` module interface design.
* Optionally preserves transparency encoding and decoding leading zeroes.
* Reasonably fast, because every library claims it is fast.
    * Faster than libraries that take generic approaches to encode any Base either via module generation or runtime alphabets
    * Faster than doing `:binary.encode_unsigned("hello") |> Integer.to_string(2)`
    * Faster than methods using power-based functions that can also overflow when using the BIF
    * Avoids the div/rem operation that is the source of many slowdowns, such as in `Integer.to_string/2` and its wrapped `:erlang.integer_to_binary/2`
* Uses less memory than most other methods
* Option to remove extra padding if losing losing leading zeroes is desirable or wanted to produce smaller resulting binaries
    * Loses transparency
* Control over shape of output binary
    * Force padding always, only for leading zeroes, or return a smaller binary representation    
    
The overall rationale for this module is:

* Avoid the aforementioned pitfalls with many popular Base2 encoding/decoding methods using BIFs only
    * Performance and memory footprint
    * Loss of leading zeroes
    * Scattered interfaces/necessity to encapsulate chained calls for consistency in a real-world app
    * Padding behavior and loss of data fidelity
* Attempt to encapsulate the fastest, most pragmatic methods for encoding and decoding Base2 in one place, with tests.
* Minimal memory footprint and good performance in tight loops or aggressive callers.
* Unified interface and consistent behavior with regard to data transparency for working with [Multibase](https://github.com/multiformats/multibase).
* Offer error handling versions of Base2 decoding.
* Consistent with the core Elixir 'Base' module.
* No extra NIFs.
* Control over padding behavior.
* Control over transparent encoding and decoding.
* Decode Base2 strings whether padded or not.

You *should* use `Base2` if:

* You want faster Base2 binary encoding than any methods like `Integer.to_string()` that are based off `div` and `rem` approaches. Both are extremely slow in Elixir.
* You need to build some sort of multi-base/base-x implementation such as Multibase](https://github.com/multiformats/multibase) where you want efficient encoding and/or decoding across different Base values.
* You want a single, unified and clear interface for working with Base2 or potentially other Base values as well.
* You want basic error handling versions for Base2 decoding without rolling it yourself
* You need specific padding behavior, perhaps for transparency or display reasons (ex: binary viewers/editors)  

You should *not* use `Base2` if:

* You do not need to represent Base2 as Elixir binaries. It is best to work with Base2 in terms of bits when possible and will be much faster than dealing with strings.
* You do not care about performance or any of the other reasons above. I probably don't want to know you either, bub.

## Usage

Encoding a simple binary:

```elixir
Base2.encode2(<<1>>)
"1"

# notice the difference
Base2.encode2("1")
"110001"

Base2.encode2("hello")
"110100001100101011011000110110001101111"
```

Encode a leading 0 binary:

```elixir
# notice the 0 is padded by default so we can decode it in exactly the same form we encoded it
Base2.encode2(<<0, 1>>)  
"0000000000000001"

# if we do not need this behavior
# note: will not decode to the same binary, instead it will be <<1>>
Base2.encode2(<<0, 1>>, padding: :none)
"1"
```

We can always pad our binary if we like, even if it does not have a leading zero:

```elixir
# always a multiple of 8
Base2.encode2(<<1>>, padding: :all) 
"00000001"

# this is quite nice if we want to do things like examine chunks uniformly, maybe for your future bin viewer/editor
for <<block::binary-8 <- Base2.encode2(<<0, 1, 2, 3>>, padding: :all)>>, do: block
["00000000", "00000001", "00000010", "00000011"]
```

Let's decode some of the things we just encoded above:

```elixir
Base2.decode2!("1")
<<1>>

Base2.decode2!("110001")
"1"

Base2.decode2!("110100001100101011011000110110001101111")
"hello"

# what if we had some leading zeroes in front of that?
# of course it transparently decodes - our input is exactly our output
Base2.decode2!("00000000110100001100101011011000110110001101111")
<<0, 104, 101, 108, 108, 111>>


# that transparently decoded, but why does it look that way?
# no worries, exactly what we wanted - it's just the leading zero, let's clip it quick and dirty to prove it
Base2.decode2!("00000000110100001100101011011000110110001101111") |> :binary.part(1, 5)
"hello"

# and let's just prove it's transparent in 1 step
Base2.encode2(<<0, "hello">>) |> Base2.decode2!()
<<0, 104, 101, 108, 108, 111>>


# what if we want error handling?
Base2.decode2("110100001100101011011000110110001101111")
{:ok, "hello"}

# Let's try something that isn't Base2
Base2.decode2("I am not a Base2 binary")
:error
```

Sometimes you might want to start with an integer. No worries, just convert it to binary using the appropriate method.

Shockingly, some people are not very familiar how to do this. Let's help them even though we're not one of those people.

```elixir
# If it's unsigned, here's a fast way:
:binary.encode_unsigned(1234) |> Base2.encode2()
"10011010010"

# If you already know the intended size, you could of course do this too
myinteger = 32767
<<myinteger::unsigned-integer-size(2)-unit(8)>> |> Base2.encode2()
"111111111111111"


# We can play with the size too easily to zero-pad it
<<myinteger::unsigned-integer-size(3)-unit(8)>> |> Base2.encode2()
"000000000111111111111111"
```

## Installation

Base2 is available via [Hex](https://hex.pm/packages/base2). The package can be installed by adding `base2` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:base2, "~> 0.1.0"}
  ]
end
```

API Documentation can be found at [https://hexdocs.pm/base2/Base2.html](https://hexdocs.pm/base2/Base2.html).


NEWS.md
# Base2 0.1.0

* Initial release

test/base2_test.exs
defmodule Base2Test do
  use ExUnit.Case
  doctest Base2

  import Base2

  test "encode/2 encodes binary as base2" do
    assert encode2(<<0>>) == "0"
    assert encode2(<<1>>) == "1"
    assert encode2(<<1, 1>>) == "100000001"
    assert encode2("hello") == "110100001100101011011000110110001101111"
  end

  test "encode/2 encodes an empty string" do
    assert encode2(<<>>) == <<>>
  end

  test "encode2/2 encodes and decodes leading zeroes" do
    assert encode2(<<0>>) == "0"
    assert encode2(<<0, 1>>) == "0000000000000001"
    assert encode2(<<0, 1, 0>>) == "000000000000000100000000"
    assert encode2(<<0, "hello">>) == "000000000110100001100101011011000110110001101111"
    assert encode2(<<0, "hello", 0>>) == "00000000011010000110010101101100011011000110111100000000"
  end

  test "transparently encodes and decodes the Base64 alphabet" do
    b64_alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
    assert encode2(b64_alphabet) |> decode2!() == b64_alphabet
    assert encode2(b64_alphabet) |> decode2() == {:ok, b64_alphabet}
  end

  test "Base2 transparently encodes binaries" do
    assert "hello" |> Base2.encode2() |> Base2.decode2!() == "hello"
    assert <<0, "hello">> |> Base2.encode2() |> Base2.decode2!() == <<0, "hello">>
  end

  test "encode/2 encodes binary as base2 with padded zeroes" do
    assert encode2(<<0>>, padding: :zeroes) == "0"
    assert encode2(<<1>>, padding: :zeroes) == "1"
    assert encode2(<<1, 1>>, padding: :zeroes) == "100000001"
    assert encode2("hello", padding: :zeroes) == "110100001100101011011000110110001101111"
  end

  test "encode/2 encodes an empty string with padded zeroes" do
    assert encode2(<<>>, padding: :zeroes) == <<>>
  end

  test "encode2/2 encodes and decodes leading zeroes with padded zeroes" do
    assert encode2(<<0>>, padding: :zeroes) == "0"
    assert encode2(<<0, 1>>, padding: :zeroes) == "0000000000000001"
    assert encode2(<<0, 1, 0>>, padding: :zeroes) == "000000000000000100000000"
    assert encode2(<<0, "hello">>, padding: :zeroes) == "000000000110100001100101011011000110110001101111"
    assert encode2(<<0, "hello", 0>>, padding: :zeroes) == "00000000011010000110010101101100011011000110111100000000"
  end

  test "transparently encodes and decodes the Base64 alphabet with padded zeroes" do
    b64_alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
    assert encode2(b64_alphabet, padding: :zeroes) |> decode2!() == b64_alphabet
    assert encode2(b64_alphabet, padding: :zeroes) |> decode2() == {:ok, b64_alphabet}
  end

  test "Base2 transparently encodes binaries with padded zeroes" do
    assert "hello" |> Base2.encode2(padding: :zeroes) |> Base2.decode2!() == "hello"
    assert <<0, "hello">> |> Base2.encode2(padding: :zeroes) |> Base2.decode2!() == <<0, "hello">>
  end

  test "encode/2 encodes binary as base2 with full padding" do
    assert encode2(<<0>>, padding: :all) == "00000000"
    assert encode2(<<1>>, padding: :all) == "00000001"
    assert encode2(<<1, 1>>, padding: :all) == "0000000100000001"
    assert encode2("hello", padding: :all) == "0110100001100101011011000110110001101111"
  end

  test "encode/2 with full padding always produces strings that are a multiple of eight in length" do
    for n <- 1..128 do
      assert :binary.encode_unsigned(n) |> Base2.encode2(padding: :all) |> String.length() |> rem(8) == 0
      assert :binary.encode_unsigned(n * 255) |> Base2.encode2(padding: :all) |> String.length() |> rem(8) == 0
    end
  end

  test "encode/2 encodes an empty string with full padding" do
    assert encode2(<<>>, padding: :all) == <<>>
  end

  test "encode2/2 encodes and decodes leading zeroes with full padding" do
    assert encode2(<<0>>, padding: :all) == "00000000"
    assert encode2(<<0, 1>>, padding: :all) == "0000000000000001"
    assert encode2(<<0, 1, 0>>, padding: :all) == "000000000000000100000000"
    assert encode2(<<0, "hello">>, padding: :all) == "000000000110100001100101011011000110110001101111"
    assert encode2(<<0, "hello", 0>>, padding: :all) == "00000000011010000110010101101100011011000110111100000000"
  end

  test "transparently encodes and decodes the Base64 alphabet with full padding" do
    b64_alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
    assert encode2(b64_alphabet, padding: :all) |> decode2!() == b64_alphabet
    assert encode2(b64_alphabet, padding: :all) |> decode2() == {:ok, b64_alphabet}
  end

  test "Base2 transparently encodes binaries with full padding" do
    assert "hello" |> Base2.encode2(padding: :all) |> Base2.decode2!() == "hello"
    assert <<0, "hello">> |> Base2.encode2(padding: :all) |> Base2.decode2!() == <<0, "hello">>
  end

  test "encode/2 encodes binary as base2 with no padding" do
    assert encode2(<<0>>, padding: :none) == "0"
    assert encode2(<<1>>, padding: :none) == "1"
    assert encode2(<<1, 1>>, padding: :none) == "100000001"
    assert encode2("hello", padding: :none) == "110100001100101011011000110110001101111"
  end

  test "encode/2 encodes an empty string with no padding" do
    assert encode2(<<>>, padding: :none) == <<>>
  end

  test "encode2/2 encodes and decodes leading zeroes with no padding" do
    assert encode2(<<0>>, padding: :none) == "0"
    assert encode2(<<0, 1>>, padding: :none) == "1"
    assert encode2(<<0, 1, 0>>, padding: :none) == "100000000"
    assert encode2(<<0, "hello">>, padding: :none) == "110100001100101011011000110110001101111"
    assert encode2(<<0, "hello", 0>>, padding: :none) == "11010000110010101101100011011000110111100000000"
  end

  test "transparently encodes and decodes the Base64 alphabet with no padding" do
    b64_alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
    assert encode2(b64_alphabet, padding: :none) |> decode2!() == b64_alphabet
    assert encode2(b64_alphabet, padding: :none) |> decode2() == {:ok, b64_alphabet}
  end

  test "Base2 transparently encodes binaries with no padding" do
    assert "hello" |> Base2.encode2(padding: :none) |> Base2.decode2!() == "hello"
    assert <<"hello world">> |> Base2.encode2(padding: :none) |> Base2.decode2!() == "hello world"
  end

  test "decode2!/1 decodes base 2 binaries" do
    assert decode2!("0") == <<0>>
    assert decode2!("1") == <<1>>
    assert decode2!("10") == <<2>>
    assert decode2!("110100001100101011011000110110001101111") == "hello"
    assert decode2!("10000000000001") == <<32, 1>>
  end

  test "decode2/1 decodes base 2 binaries" do
    assert decode2("0") == {:ok, <<0>>}
    assert decode2("1") == {:ok, <<1>>}
    assert decode2("10") == {:ok, <<2>>}
    assert decode2("110100001100101011011000110110001101111") == {:ok, "hello"}
    assert decode2("10000000000001") == {:ok, <<32, 1>>}
  end

  test "decode2!/1 decodes base 2 binaries with 0 padding" do
    assert decode2!("000000000") == <<0, 0>>
    assert decode2!("000000001") == <<0, 1>>
    assert decode2!("0000000010") == <<0, 2>>
    assert decode2!("000000000000000010") == <<0, 0, 2>>
  end

  test "decode2/1 decodes base 2 binaries with 0 padding" do
    assert decode2("000000000") == {:ok, <<0, 0>>}
    assert decode2("000000001") == {:ok, <<0, 1>>}
    assert decode2("0000000010") == {:ok, <<0, 2>>}
    assert decode2("000000000000000010") == {:ok, <<0, 0, 2>>}
  end

  test "decode2/1 handles invalid input" do
    assert "I am not Base2 even if I want to be" |> decode2() == :error
    assert "101010101COMPUTERWELT101010101001" |> decode2() == :error
    assert <<32, "1010100101">> |> decode2() == :error
    assert <<"101010101010", 32>> |> decode2() == :error
    assert "-1" |> decode2() == :error
  end

  test "decode2!/1 raises an error for invalid input" do
    assert_raise ArgumentError, fn ->  "I am not Base2 even if I want to be" |> decode2!() end
    assert_raise ArgumentError, fn -> "101010101COMPUTERWELT101010101001" |> decode2!() end
    assert_raise ArgumentError, fn -> <<32, "1010100101">> |> decode2!() end
    assert_raise ArgumentError, fn -> <<"101010101010", 32>> |> decode2!() end
    assert_raise ArgumentError, fn -> "-1" |> decode2!() end
  end

end


mix.exs
defmodule Base2.MixProject do
  use Mix.Project

  @version "0.1.0"
  @source_url "https://github.com/nocursor/base2"

  def project do
    [
      app: :base2,
      version: @version,
      elixir: ">= 1.7.0",
      start_permanent: Mix.env() == :prod,
      deps: deps(),

      # Hex
      description: "Elixir library for encoding and decoding binaries in Base2.",
      package: package(),

      # Docs
      name: "Base2",
      source_url: @source_url,
      docs: docs()
    ]
  end

  def application do
    [
    ]
  end

  defp package do
    [
      maintainers: [
        "nocursor",
      ],
      licenses: ["MIT"],
      links: %{github: @source_url},
      files: ~w(lib NEWS.md LICENSE.md mix.exs README.md)
    ]
  end

  defp docs do
    [
      main: "readme",
      source_ref: "v#{@version}",
      extra_section: "PAGES",
      extras: extras(),
    ]
  end

  defp extras do
    [
      "README.md",
    ]
  end

  defp deps do
    [
      {:benchee, "~> 0.13.2", only: [:dev]},
      {:dialyxir, "~> 1.0.0-rc.3", only: [:dev], runtime: false},
      {:ex_doc, "~> 0.13", only: [:dev], runtime: false},
    ]
  end
end


lib/base2.ex
defmodule Base2 do
  @moduledoc """
  This module provides data encoding and decoding functions for Base2.

  ## Overview

  Converting to and from Base2 is very simple in Elixir and Erlang. Unfortunately, most approaches use generic methodologies that are suitable for any Base, and thus do not optimize for any Base typically.

  Working with Base2 is a relatively simple task chaining a few Elixir built-in functions or using a third-party generic "BaseX" type library, but most of these implementations leave a lot to be desired. Generally, most built-in methods and third-party libraries are often not very optimized. Using built-in functions also is not uniform with other ways of handling Bases such as via the Elixir `Base` module. Most of these methods are great for scratch work, but less suitable for bulk encoding and decoding.  Further, the multiple ways of approaching different bases lead to very inconsistent interfaces, for instance `Integer.to_string()` vs. `Base` vs a third-party module with its own conventions.

  `Base2` includes the following functionality:

  * Encodes and Decodes binaries as Base2.
  * Consistent with the `Base` module interface design.
  * Optionally preserves transparency encoding and decoding leading zeroes.
  * Reasonably fast, because every library claims it is fast.
    * Faster than libraries that take generic approaches to encode any Base either via module generation or runtime alphabets
    * Faster than doing `:binary.encode_unsigned("hello") |> Integer.to_string(2)`
    * Faster than methods using power-based functions that can also overflow when using the BIF
    * Avoids the div/rem operation that is the source of many slowdowns, such as in `Integer.to_string/2` and its wrapped `:erlang.integer_to_binary/2`
  * Uses less memory than most other methods
  * Option to remove extra padding if losing losing leading zeroes is desirable or wanted to produce smaller resulting binaries
    * Loses transparency
  * Control over shape of output binary
    * Force padding always, only for leading zeroes, or return a smaller binary representation

  ## Padding

  `Base2` allows full control over padding behavior when encoding a binary into a Base2 string.

  There are three options for padding:

    * `:zeroes` (default) - Allows zeroes to be padded to ensure transparent decode of leading zeroes.
    * `:all` - Uniformly pads the data with zeroes. String length will always a multiple of 8 and fast, but at the cost of an increased output size.
    * `:none` - Produces a smaller representation by dropping all leading zeroes, but at the cost of fully transparent decode if there are leading zeroes.

    `:zeroes` is good for general usage, while typically being smaller. Zeroes are fully padded to byte boundaries. Transparency is fully preserved.

    `:all` is good for uniformly working with Base2 string output. Transparency is fully preserved.

    `:none` exhibit the same behavior as methods such as `Integer.to_string("hello", 2)` that try to use a smaller representation. This method should be used with caution as it comes at a cost of transparency. Simply, if you want the exact same output when decoding as your input, do not use this option. If, however, you want the smallest representation, then it is a good choice.

  """

  @typedoc """
  A Base2 encoded string.
  """
  @type base2_binary() :: binary()

  @doc """
  Encodes a binary string into a base 2 encoded string.

  Accepts a `:padding` option which will control the padding behavior of the resulting strings.

  The options for `:padding` can be:

    * `:zeroes` (default) - Allows zeroes to be padded to ensure transparent decode of leading zeroes.
    * `:all` - Uniformly pads the data with zeroes. String length will always a multiple of 8 and fast, but at the cost of an increased output size.
    * `:none` - Produces a smaller representation by dropping all leading zeroes, but at the cost of fully transparent decode if there are leading zeroes.

  ## Examples

      iex> Base2.encode2("hello world")
      "110100001100101011011000110110001101111001000000111011101101111011100100110110001100100"

      iex> Base2.encode2(<<0, 1>>, padding: :zeroes)
      "0000000000000001"

      iex> Base2.encode2(<<0, 1>>, padding: :none)
      "1"

      iex> Base2.encode2(<<0, 1>>, padding: :all)
      "0000000000000001"

      iex> Base2.encode2(<<1>>, padding: :all)
      "00000001"

  """
  @spec encode2(binary(), keyword()) :: base2_binary()
  def encode2(binary, opts \\ [])
  def encode2(binary, opts) when is_binary(binary) do
    pad_type = Keyword.get(opts, :padding, :zeroes)
    do_encode(binary, pad_type)
  end

  @doc """
  Decodes a base 2 encoded string as a binary string.

  An ArgumentError is raised if the string is not a base 2 encoded string.

  ## Examples

      iex> Base2.decode2!("110100001100101011011000110110001101111001000000111011101101111011100100110110001100100")
      "hello world"

      iex> Base2.decode2!("1")
      <<1>>

      iex> Base2.decode2!("0000000000000001")
      <<0, 1>>

      iex> Base2.decode2!("00000001")
      <<1>>

  """
  @spec decode2!(base2_binary()) :: binary()
  def decode2!(string) when is_binary(string) do
    do_decode(string)
  end

  @doc """
  Decodes a base 2 encoded string as a binary string.

  Returns `:error` if the string is not a base 2 encoded string.

  ## Examples

      iex> Base2.decode2("110100001100101011011000110110001101111001000000111011101101111011100100110110001100100")
      {:ok, "hello world"}

      iex> Base2.decode2("1")
      {:ok, <<1>>}

      iex> Base2.decode2("0000000000000001")
      {:ok, <<0, 1>>}

      iex> Base2.decode2("00000001")
      {:ok, <<1>>}

      iex> Base2.decode2("hello world")
      :error

      iex> Base2.decode2("00101015")
      :error

  """
  @spec decode2(base2_binary()) :: {:ok, binary()} | :error
  def decode2(string) when is_binary(string) do
    {:ok, decode2!(string)}
    rescue
      ArgumentError -> :error
  end

#===============================================================================
# Private Island
#===============================================================================

  defp do_encode(data, :none) do
    trim_encode(data, false)
  end

  defp do_encode(data, :zeroes) do
    trim_encode(data, true)
  end

  defp do_encode(data, :all) do
    encode_body(data, [])
  end

  defp trim_encode(<<>>, _pad?) do
    <<>>
  end

  defp trim_encode(<<0>>, _pad?) do
    "0"
  end

  defp trim_encode(data, pad?) do
    trim_encode(data, [], pad?)
  end

  defp trim_encode(<<0, rest::binary>>, _acc, true) do
    # here we could try to make a more compact representation, but it complicates decoding so not sure if it's worth it
    encode_body(rest, '00000000')
  end

  defp trim_encode(<<1::1, rest::bitstring>>, _acc, _pad?) do
    encode_body(rest, ['1'])
  end

  defp trim_encode(<<0::1, rest::bitstring>>, acc, pad?) do
    trim_encode(rest, acc, pad?)
  end

  defp trim_encode(<<>>, [], _pad?) do
    <<>>
  end

  defp encode_body(<<>>, acc) do
    acc |> Enum.reverse() |> to_string()
  end

  defp encode_body(<<1::1, rest::bitstring>>, acc) do
    encode_body(rest, ['1' | acc])
  end

  defp encode_body(<<0::1, rest::bitstring>>, acc) do
    encode_body(rest, ['0' | acc])
  end

  defp count_leading_zeroes(<<"00000000", rest::binary>>, acc) do
    # This call here could be tweaked with another few pattern matches if we want a smaller representation for the case of `:zeroes` padding, however experimenting, the decoding speed was a bit less
    count_leading_zeroes(rest, acc + 1)
  end

  defp count_leading_zeroes(_string, acc) do
    acc
  end

  defp do_decode(string) do
    # Using String.to_integer + encode_unsigned is much faster than manual recursive decode to build individual bits
    leading_zeroes = count_leading_zeroes(string, 0)
    <<0::size(leading_zeroes)-unit(8), (String.to_integer(string, 2) |> :binary.encode_unsigned())::binary>>
  end

end
